File: src/coldchain/api/schemas.py
"""Typed HTTP request and response models for Milestone 1A."""

from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class ApiModel(BaseModel):
    model_config = ConfigDict(extra="forbid", alias_generator=to_camel, populate_by_name=True)


class TelemetryAccepted(ApiModel):
    event_id: UUID
    status: str


class TelemetryReadingView(ApiModel):
    event_id: UUID
    occurred_at: str
    reading_sequence: int
    temperature_celsius: float
    cargo_type: str
    disposition: str
    policy_inputs: dict[str, Any]
    reason_codes: list[str]


class EvidenceSnapshotView(ApiModel):
    evidence_id: UUID
    telemetry_event_ids: list[UUID]
    policy_inputs: dict[str, Any]
    summary: str
    content_hash: str
    captured_at: str


class RecommendationDetailView(ApiModel):
    recommendation_id: UUID
    evidence_ids: list[UUID]
    action_type: str
    parameters: dict[str, Any]
    provider: str
    author_identity: str
    rationale: str
    expires_at: str
    non_authoritative: bool
    provenance: dict[str, Any]


class GovernanceEvaluationView(ApiModel):
    evaluation_id: UUID
    recommendation_id: UUID
    decision: str
    policy_version: str
    reason_codes: list[str]
    inputs: dict[str, Any]
    evaluated_at: str


class ApprovalView(ApiModel):
    approval_id: UUID
    recommendation_id: UUID
    actor_id: str
    actor_role: str
    decision: str
    rationale: str
    decided_at: str


class IncidentView(ApiModel):
    incident_id: UUID
    shipment_id: UUID
    state: str
    severity: str
    policy_version: str
    source_event_ids: list[UUID]
    correlation_id: UUID
    created_at: str
    updated_at: str
    recommendation_id: UUID | None = None
    recommended_action: str | None = None
    recommendation_expires_at: str | None = None
    recommendation_provenance: dict[str, Any] | None = None
    telemetry: list[TelemetryReadingView] = Field(default_factory=list)
    evidence: EvidenceSnapshotView | None = None
    recommendation: RecommendationDetailView | None = None
    governance: GovernanceEvaluationView | None = None
    approval: ApprovalView | None = None
    command: "CommandStatusView | None" = None


class TimelineEventView(ApiModel):
    sequence: int
    audit_event_id: UUID
    event_type: str
    actor_id: str
    correlation_id: UUID
    causation_id: UUID | None
    component_version: str
    schema_version: str
    payload: dict[str, Any]
    occurred_at: str
    previous_event_hash: str | None
    event_hash: str


class DecisionRequest(ApiModel):
    rationale: str = Field(min_length=1, max_length=2000)
    idempotency_key: str = Field(min_length=16, max_length=200)


class ActionResultView(ApiModel):
    result_id: UUID
    status: str
    adapter: str
    detail: str
    completed_at: str


class DecisionResponse(ApiModel):
    approval_id: UUID
    decision: str
    idempotent_replay: bool
    command_id: UUID | None = None
    status: str | None = None
    action_result: ActionResultView | None = None


class CommandStatusView(ApiModel):
    command_id: UUID
    incident_id: UUID
    action_type: str
    status: str
    adapter: str | None
    detail: str | None
    action_result: ActionResultView | None = None


class DemoScenarioView(ApiModel):
    scenario_id: str
    title: str
    description: str
    expected_outcome: str
    human_action: str


class DemoRunStarted(ApiModel):
    run_id: UUID
    scenario_id: str
    correlation_id: UUID
    shipment_id: UUID
    event_ids: list[UUID]
    publish_count: int
    status: str


class DemoRunSteps(ApiModel):
    telemetry_submitted: bool
    event_accepted: bool
    policy_evaluated: bool
    evidence_collected: bool
    recommendation_created: bool
    governance_completed: bool
    incident_ready: bool


class DemoRunStatus(ApiModel):
    run_id: UUID
    scenario_id: str
    correlation_id: UUID
    shipment_id: UUID
    event_ids: list[UUID]
    publish_count: int
    started_at: str
    status: str
    processed_event_count: int
    dispositions: list[str]
    incident: IncidentView | None
    steps: DemoRunSteps


IncidentView.model_rebuild()

File: frontend/src/data/api/schemas.ts
import { z } from 'zod';

const uuid = z.string().uuid();
const record = z.record(z.string(), z.unknown());
const object = z.strictObject;

export const telemetrySchema = object({ eventId: uuid, occurredAt: z.string(), readingSequence: z.number().int(), temperatureCelsius: z.number(), cargoType: z.string(), disposition: z.string(), policyInputs: record, reasonCodes: z.array(z.string()) });
export const evidenceSchema = object({ evidenceId: uuid, telemetryEventIds: z.array(uuid), policyInputs: record, summary: z.string(), contentHash: z.string(), capturedAt: z.string() });
export const recommendationSchema = object({ recommendationId: uuid, evidenceIds: z.array(uuid), actionType: z.string(), parameters: record, provider: z.string(), authorIdentity: z.string(), rationale: z.string(), expiresAt: z.string(), nonAuthoritative: z.boolean(), provenance: record });
export const governanceSchema = object({ evaluationId: uuid, recommendationId: uuid, decision: z.string(), policyVersion: z.string(), reasonCodes: z.array(z.string()), inputs: record, evaluatedAt: z.string() });
export const approvalSchema = object({ approvalId: uuid, recommendationId: uuid, actorId: z.string(), actorRole: z.string(), decision: z.string(), rationale: z.string(), decidedAt: z.string() });
export const actionResultSchema = object({ resultId: uuid, status: z.string(), adapter: z.string(), detail: z.string(), completedAt: z.string() });
export const commandSchema = object({ commandId: uuid, incidentId: uuid, actionType: z.string(), status: z.string(), adapter: z.string().nullable(), detail: z.string().nullable(), actionResult: actionResultSchema.nullable().optional() });
export const incidentSchema = object({ incidentId: uuid, shipmentId: uuid, state: z.string(), severity: z.string(), policyVersion: z.string(), sourceEventIds: z.array(uuid), correlationId: uuid, createdAt: z.string(), updatedAt: z.string(), recommendationId: uuid.nullable().optional(), recommendedAction: z.string().nullable().optional(), recommendationExpiresAt: z.string().nullable().optional(), recommendationProvenance: record.nullable().optional(), telemetry: z.array(telemetrySchema).default([]), evidence: evidenceSchema.nullable().optional(), recommendation: recommendationSchema.nullable().optional(), governance: governanceSchema.nullable().optional(), approval: approvalSchema.nullable().optional(), command: commandSchema.nullable().optional() });
export const incidentsSchema = z.array(incidentSchema);
export const timelineSchema = z.array(object({ sequence: z.number().int(), auditEventId: uuid, eventType: z.string(), actorId: z.string(), correlationId: uuid, causationId: uuid.nullable(), componentVersion: z.string(), schemaVersion: z.string(), payload: record, occurredAt: z.string(), previousEventHash: z.string().nullable(), eventHash: z.string() }));
export const decisionSchema = object({ approvalId: uuid, decision: z.enum(['APPROVED', 'REJECTED']), idempotentReplay: z.boolean(), commandId: uuid.nullable().optional(), status: z.string().nullable().optional(), actionResult: actionResultSchema.nullable().optional() });
export const scenarioSchema = object({ scenarioId: z.string(), title: z.string(), description: z.string(), expectedOutcome: z.string(), humanAction: z.string() });
export const scenariosSchema = z.array(scenarioSchema);
export const demoRunSchema = object({ runId: uuid, scenarioId: z.string(), correlationId: uuid, shipmentId: uuid, eventIds: z.array(uuid), publishCount: z.number().int(), startedAt: z.string().optional(), status: z.enum(['ACCEPTED', 'PROCESSING', 'INCIDENT_READY', 'COMPLETED_WITHOUT_INCIDENT']), processedEventCount: z.number().int().default(0), dispositions: z.array(z.string()).default([]), incident: incidentSchema.nullable().optional(), steps: object({ telemetrySubmitted: z.boolean(), eventAccepted: z.boolean(), policyEvaluated: z.boolean(), evidenceCollected: z.boolean(), recommendationCreated: z.boolean(), governanceCompleted: z.boolean(), incidentReady: z.boolean() }).default({ telemetrySubmitted: true, eventAccepted: true, policyEvaluated: false, evidenceCollected: false, recommendationCreated: false, governanceCompleted: false, incidentReady: false }) });
export const liveHealthSchema = object({ status: z.literal('alive') });
export const aiHealthSchema = object({ selected_provider: z.string(), selected_model: z.string(), provider_configured: z.boolean(), live_calls_enabled: z.boolean(), qdrant_readiness: z.string(), embedding_provider_readiness: z.string(), fallback_available: z.boolean() });

export type ApiIncident = z.infer<typeof incidentSchema>;
export type ApiTimeline = z.infer<typeof timelineSchema>;
export type ApiDecision = z.infer<typeof decisionSchema>;
export type ApiDemoRun = z.infer<typeof demoRunSchema>;
